src/funnels/Swapper.sol

```diff
@@ -0,0 +1,87 @@
+// SPDX-FileCopyrightText: © 2021 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.16;
+
+interface RolesLike {
+    function canCall(bytes32, address, address, bytes4) external view returns (bool);
+}
+
+interface GemLike {
+    function balanceOf(address) external view returns (uint256);
+    function transferFrom(address, address, uint256) external;
+}
+
+interface CalleeLike {
+    function swap(address src, address dst, uint256 amt, uint256 minOut, address to, bytes calldata data) external;
+}
+
+contract Swapper {
+    mapping (address => uint256) public wards;
+    mapping (address => mapping (address => uint256)) public hops; // [seconds]  hops[src][dst] is the swap cooldown when swapping `src` to `dst`.
+    mapping (address => mapping (address => uint256)) public zzz;  // [seconds]   zzz[src][dst] is the timestamp of the last swap from `src` to `dst`.
+    mapping (address => mapping (address => uint256)) public caps; // [weis]     caps[src][dst] is the maximum amount that can be swapped each hop when swapping `src` to `dst`.
+
+
+    address   public immutable buffer;                // Contract from which the GEM to sell is pulled and to which the bought GEM is pushed
+    RolesLike public immutable roles;                 // Contract managing access control for this Depositor
+    bytes32   public immutable ilk;
+
+    event Rely (address indexed usr);
+    event Deny (address indexed usr);
+    event File (bytes32 indexed what, address indexed src, address indexed dst, uint256 data);
+    event Swap (address indexed sender, address indexed src, address indexed dst, uint256 amt, uint256 out);
+
+    constructor(address roles_, bytes32 ilk_, address buffer_) {
+        roles = RolesLike(roles_);
+        ilk = ilk_;
+        buffer = buffer_;
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth() {
+        require(roles.canCall(ilk, msg.sender, address(this), msg.sig) || wards[msg.sender] == 1, "Swapper/not-authorized");
+        _;
+    }
+
+    function rely(address usr) external auth { wards[usr] = 1; emit Rely(usr); }
+    function deny(address usr) external auth { wards[usr] = 0; emit Deny(usr); }
```

Comment:
Nit: I prefer the standard format (instead of the one line one), but not too strong opinion either.

Reply:
Well we are removing one pair of them.

Reply:
yeah, I guess if it's 2 pairs I don't mind either way.

Reply:
I feel it is kind of anti-pattern to treat them differently, even if the three pairs would remain, but taking into account that we are going with just 2 let's leave that discussion for a next time :)